src/lib/client.ts

```diff
@@ -512,7 +516,7 @@ export default class MqttClient extends TypedEventEmitter<MqttClientEventCallbac
 		} else {
 			this.log(
 				'MqttClient :: environment',
-				isBrowser ? 'browser' : 'node',
+				options.forceNativeWebSocket || isBrowser ? 'browser' : 'node',
```

Comment:
instead of having this `||` codewise I would make isBrowser a function that accepts client options as argument (the argument can be optional and if not present it will simply return the isBrowser value)

Reply:
I am wondering then if this make sense that this specific variable define the env globally. I meant, it makes sense to use that specific variable to determine which websocket lib to use, but that's it. Even if, we could then determine the env just based on that choice. Maybe the name of `forceNativeWebSocket` was a bad idea from my side.

Same for:
https://github.com/mqttjs/MQTT.js/blob/6b278dca5a5b82b07835344f3c129ddd5b73e6e8/src/lib/get-timer.ts#L23-L39
`get-timer` also use `isBrowser`, should I also rely on the client's options, namely `forceNativeWebSocket`? I could instead pass to the function all the clients options instead of just the `timerVariant` value.

Btw, if we use `forceNativeWebSocket` inside `isBrowser`, then `timeVariant` may also be used inside `isBrowser`.

So in the end, we could have instead an `environment: "auto" | "browser" | "node" | "react-native | "electron"` option as proposed in #1900, give it to `isBrowser` or nothing for "auto". 
This will give the same result, but using a more generic approach, e.g. we could mark as deprecated `timerVariant`. 
However, this will give less flexibility, because if you are under a specific env, but you do not want the choice imposed by us for the underling libs used; I am mainly thinking about a recent node env that want to use native WebSocket.

--- 
1. Should I generalize to an `environment` option?
2. Should I use `forceNativeWebSocket` *only* for websocket choice related, thus not inside `isBrowser` (and in the snippet of your review)?
3. `environment` + `forceNativeWebSocket` options?
4. Also use `forceNativeWebSocket` to influence `get-timer`, or use `environment`?
5. ?

Reply:
Giving that we already have an option to allow users specifically choose the timer variant IMO it doens't make so much sense here to use. To avoid confusion we could keep exporting the `isBrowser` const and also export a `shouldUseNativeWs` function with the client options as argument as said before.

Using environment is not a good idea as it doens't work well in browsers so I think your solution works better in this case.

So I would go for solution 2